src/linear_algebra/alternating.lean

```diff
@@ -434,3 +444,292 @@ lemma comp_multilinear_map_alternatization (g : N' →ₗ[R] N'₂)
+
+end equiv.perm
+
+namespace alternating_map
+open equiv
+
+/-- Auxiliary helper for `dom_coprod`, before it has been proven that the result is alternating. -/
+def dom_coprod_fun
+  {R : Type*} {M N₁ N₂ : Type*}
+  [comm_semiring R]
+  [add_comm_group N₁] [semimodule R N₁]
+  [add_comm_group N₂] [semimodule R N₂]
+  [add_comm_monoid M] [semimodule R M]
+  (a : alternating_map R M N₁ ιa) (b : alternating_map R M N₂ ιb) :
+  multilinear_map R (λ _ : ιa ⊕ ιb, M) (N₁ ⊗[R] N₂) :=
+∑ σ : perm.mod_sum_congr ιa ιb, σ.lift_on' (λ σ,
+  (σ.sign : ℤ) •
+    (multilinear_map.dom_coprod a b : multilinear_map R (λ (_ : ιa ⊕ ιb), M) (N₁ ⊗ N₂))
+      .dom_dom_congr σ)
+(λ σ₁ σ₂ h, begin
+  ext v,
+  simp only [multilinear_map.dom_dom_congr_apply, multilinear_map.dom_coprod_apply,
+    coe_multilinear_map, multilinear_map.smul_apply],
+  obtain ⟨⟨sl, sr⟩, h⟩ := h,
+  replace h := inv_mul_eq_iff_eq_mul.mp h.symm,
+  have : ((σ₁ * perm.sum_congr_hom _ _ (sl, sr)).sign : ℤ) = σ₁.sign * (sl.sign * sr.sign) :=
+    by simp,
+  rw [h, this, mul_smul, mul_smul, units.smul_left_cancel,
+    ←tensor_product.tmul_smul, tensor_product.smul_tmul'],
+  simp only [sum.map_inr, perm.sum_congr_hom_apply, perm.sum_congr_apply, sum.map_inl,
+             function.comp_app, perm.coe_mul],
+  rw [←a.map_congr_perm (λ i, v (σ₁ _)), ←b.map_congr_perm (λ i, v (σ₁ _))],
+end)
+
+private lemma dom_coprod_fun_eq_zero_if_eq
+  {R : Type*} {M N₁ N₂ : Type*}
+  [comm_semiring R]
+  [add_comm_group N₁] [semimodule R N₁]
+  [add_comm_group N₂] [semimodule R N₂]
+  [add_comm_monoid M] [semimodule R M]
+  (a : alternating_map R M N₁ ιa) (b : alternating_map R M N₂ ιb)
+  (v : ιa ⊕ ιb → M) (i j : ιa ⊕ ιb) (hv : v i = v j) (hij : i ≠ j) :
+  dom_coprod_fun a b v = 0 :=
+begin
+  unfold dom_coprod_fun,
+  dsimp only,
+  rw multilinear_map.sum_apply,
+  apply finset.sum_involution
+    (λ σ _, (equiv.swap i j • σ : perm.mod_sum_congr ιa ιb))
+    (λ σ, _)
+    (λ σ, _)
+    (λ σ _, finset.mem_univ _)
+    (λ σ, _),
+  all_goals {
+    apply σ.induction_on' (λ σ, _),
+    rintro _, },
+  { dsimp only [quotient.lift_on'_mk', quotient.map'_mk', mul_action.quotient.smul_mk],
+    rw [perm.sign_mul, perm.sign_swap hij],
+    simp only [one_mul, units.neg_mul, function.comp_app, neg_smul, perm.coe_mul,
+      units.coe_neg, multilinear_map.smul_apply, multilinear_map.neg_apply,
+      multilinear_map.dom_dom_congr_apply, multilinear_map.dom_coprod_apply],
+    convert add_right_neg _;
+    { ext k, rw equiv.apply_swap_eq_self hv }, },
+  { dsimp only [quotient.lift_on'_mk', quotient.map'_mk', multilinear_map.smul_apply,
+      multilinear_map.dom_dom_congr_apply, multilinear_map.dom_coprod_apply],
+    apply mt,
+    intro hσ,
+    with_cases {
+      cases hi : σ⁻¹ i;
+        cases hj : σ⁻¹ j;
+        rw perm.inv_eq_iff_eq at hi hj;
+        propagate_tags' { substs hi hj, }, },
+    case [sum.inl sum.inr : i' j', sum.inr sum.inl : i' j'] {
```

Comment:
@gebner, this is where I needed `propagate_tags` in leanprover-community/lean#540. Can you think of a better solution? Should `substs` just call `propagate_tags` itself?

Reply:
Good lord! I think `propagate_tags` is the least of this proof's problems. Is there any way it can be split up into more manageable and maybe even independently useful chunks?

But yes, `substs` should call `propagate_tags` itself.

Reply:
Sorry for the first impressions.  I don't find it particularly readable when every line is either `all_goals` or `work_on_goal`.  While I value avoiding duplication, I think it would be better to refactor `all_goals { a }, work_on_goal 0 { b }, work_on_goal 1 { b' }, all_goals { c }` into `{ a, b, c }, { a, b', c }` here.

I would also introduce a definition for the summand of `dom_coprod_fun`.  This is an unwieldy expression, and you can probably extract two of the big blocks in the proof here.

Reply:
> I would also introduce a definition for the summand of dom_coprod_fun. This is an unwieldy expression, and you can probably extract two of the big blocks in the proof here.

I've done this as `dom_coprod.summand` and eliminated `dom_coprod_fun` - good idea.